<h1 align="center">Física Computacional.</h1>
<h1 align="center">Semestre 2027-1</h1>

<h2>Sergio A. Alcalá Corona </h2> 

---
### Rodrigo Vega Vilchis
### Joel Gómez Uribe
---

<h1 align="center">Programación para la física computacional</h1> 

## Práctica 1.  Programación básica para la física
---
<h2>Diaz Castro Gerson Joshua </h2> 
---

### **EJERCICIOS**:

$\;$

### **1. Coordenadas polares y cartesianas:** 
#### Supongamos que se nos da (es decir, el usuario introduce) la posición de un punto en un espacio bidimensional:

#### **a)** Si el punto se especifica  en coordenadas polares $r$,$\theta$. Escribe un programa (en `python` sin usar la biblioteca `numpy`) que calcule e imprima la conversión a coordenadas cartesianas $x$, $y$.

Solución:
Primero le pedimos al programa dos puntos cualesquiera en el plano cartesiano, y los guardamos en dos variables, $x$ y $y$:

In [1]:
x = float(input("Coloque su coordenada x: "))
y = float(input("Coloque su coordenada y: "))

Coloque su coordenada x: 8
Coloque su coordenada y: 4


Ahora recordemos que para cualquier punto, la transformación de coordenadas cartesianas a polares está dada por:$$r = \sqrt{x^2 + y^2}$$$$\theta = \arctan\left(\frac{y}{x}\right)$$

Vamos a necesitar la función arcotangente; para ello, importemos la librería math:

In [2]:
import math as mt

Ahora si podemos programar la transformacion:

In [3]:
 #Que en codigo seria:
r = ((x**2)+(y**2))**(1/2)
θ = mt.atan(y/x)

Por ultimo imprimimos los valores de $r$ y de $\theta$:

In [4]:
print("Mis coordenadas en polares son: ")
print(f"r = {r} y θ = {θ} rad")

Mis coordenadas en polares son: 
r = 8.94427190999916 y θ = 0.4636476090008061 rad


Si quisiéramos el resultado en grados, usamos la función mt.degrees para convertir de radianes a grados:

In [5]:
print("Mis coordenadas en polares son: ")
print(f"r = {r} y θ = {mt.degrees(θ)}° ")

Mis coordenadas en polares son: 
r = 8.94427190999916 y θ = 26.56505117707799° 


#### **b)** Si el punto se especifica en coordenadas cartesianas $x$, $y$. Escribe un programa (en `python` sin usar `numpy`) que calcule e imprima la conversión a coordenadas polares $r$,$\theta$.

Ahora recordemos que para cualquier punto, la transformación a coordenadas cartesianas es:$$x = r \cos(\theta)$$$$y = r \sin(\theta)$$

In [6]:
#Por lo tanto:
X = r*mt.cos(θ)
Y = r*mt.sin(θ)

Al imprimir los valores de X y Y deberian salir nuestras coordenadas originales:

In [7]:
print("Mis coordenadas originales son: ")
print(f"x = {X} y Y = {Y}")

Mis coordenadas originales son: 
x = 8.0 y Y = 4.0


### **2. Caída de una pelota desde una torre:** 
#### Se deja caer una pelota desde una torre de altura $h$. Tiene velocidad inicial cero y acelera hacia abajo con la gravedad.

Escribe un programa que le pida al usuario que ingrese la altura en metros de la torre y luego calcule e imprima el tiempo en segundos hasta que la pelota toque el suelo (ignora la resistencia del aire).

Después, usa tu programa para calcular el tiempo de una pelota lanzada desde una torre de $100m$ de altura.\\


Recordemos que la ecuación de caída libre es: $$y(t) = y_{0} + v_{o} t - \frac{1}{2}gt^{2}$$

Pero:
$$v_{0} = 0$$ $$y(t) = 0$$ $$y_{0} = h$$


Despejando t: $$t = \sqrt{\frac{2h}{g}}$$

In [16]:
#EL codigo seria:
#Definimos nuestro valor de h y de g

g = 9.81

#Colocamos un while para que nuestro programa calcule lo que buscamos y para asegurar que solo se usen valores positivos

while (True):
    h = float(input("Coloque la altura del edificio en metros: "))
    if (h > 0):
        t = mt.sqrt((2*h)/g)
        print(f"El tiempo que tarda en caer el objeto es: {t} segundos")
        break #Para que se detenga y no itere infinitamente
    elif (h <= 0):
        print("Error: La altura debe ser mayor a 0.")
        

Coloque la altura del edificio en metros: 0
Error: La altura debe ser mayor a 0.
Coloque la altura del edificio en metros: 100
El tiempo que tarda en caer el objeto es: 4.515236409857309 segundos


Colocando en mi programa el valor de $100$, el tiempo que tardaría la pelota en caer sería: $$t = \sqrt{\frac{2(100)}{9.81}}$$$$t \approx 4.515 \text{ s}$$

### **3. Relatividad especial:** 
#### Una nave espacial viaja desde la Tierra en línea recta a una velocidad relativista $v$ a otro planeta a $x$ años luz de distancia. 

#### Escribe un programa que le pida al usuario el valor de $x$ y la velocidad $v$ como una fracción de la velocidad de la luz $c$, y que imprima el tiempo en años que tarda la nave espacial en llegar a su destino

#### **(a)** en el marco de reposo de un observador en la Tierra y


En el marco terrestre, la distancia a recorrer es $x$ y el tiempo de viaje $t$ a velocidad constante está dado por: $$t = \frac{x}{v}$$

Sustituyendo nuestros datos:$$t = \frac{10*c años}{0.99*c}$$

In [17]:
#Entonces el programa seria:
while (True):
    v = float(input("Ingrese la velocidad como fracción de c (valor entre 0 y 1): "))
    x = float(input("Ingrese en años luz la distancia: "))
    if ((0 < v < 1) and (x > 0)):
        t = x/v
        print(f"El tiempo que tarda desde el marco de refrecia de la tierra es: {t} años")
        break
    else:
        print("Valores no validos, intente de nuevo:")

Ingrese la velocidad como fracción de c (valor entre 0 y 1): 4
Ingrese en años luz la distancia: 10
Valores no validos, intente de nuevo:
Ingrese la velocidad como fracción de c (valor entre 0 y 1): 0.99
Ingrese en años luz la distancia: 10
El tiempo que tarda desde el marco de refrecia de la tierra es: 10.1010101010101 años


#### **(b)** como lo percibiría un pasajero a bordo de la nave. 

Por otro lado, el pasajero experimenta un tiempo propio $\tau$ debido a que su marco de referencia es inercial; por lo tanto: $$t = \gamma \tau$$

Donde $\gamma$ es el factor de Lorentz: $$\gamma = \sqrt{1-(\frac{v}{c})^{2}}$$

Despejando a $\tau$: $$\tau = t \sqrt{1-(\frac{v}{c})^{2}}$$

In [18]:
#EL programa seria:
#Como el tiempo esta en años luz se cancela la c dentro de la raiz
τ = t*(mt.sqrt(1-(v**2)))
print(f"El tiempo que tarda desde el marco de referencia de la nave es: {τ} años")

El tiempo que tarda desde el marco de referencia de la nave es: 1.424922826228878 años


#### Usa tu programa para calcular las respuestas para un planeta a 10 años luz de distancia con $v = 0.99c$

Por lo tanto, si el planeta está a 10 años luz y su velocidad es de $0.99c$, el tiempo que tarda en llegar desde el marco de referencia de la Tierra sería: $$t = 10.101 años$$

Asi, el tiempo que tarda desde el marco de referencia de la nave es: $$ \tau = 1.424 años$$

### **4. Paso de potencial cuántico.** 
#### Un conocido problema de mecánica cuántica involucra una partícula de masa $m$ que encuentra un paso de potencial unidimensional, como este:

<center><img src="qstep2.png" width="50%" class="center"/></center>

#### La partícula con energía cinética inicial $E$ y vector de onda $k_1=\sqrt{2mE}/\hbar$ entra por la izquierda y encuentra un salto repentino en la energía potencial de altura $V$ en la posición $x=0$. Resolviendo la ecuación de Schrödinger, se puede demostrar que cuando $E>V$ la partícula puede 

#### (a) **pasar el escalón**, en cuyo caso tiene una energía cinética menor a $E-V$ en el otro lado y un vector de onda correspondientemente menor a $k_2=\sqrt{2m(E-V)}/\hbar$, \; o bien (b) **reflejarse**, manteniendo toda su energía cinética y un vector de onda sin cambios, pero moviéndose en la dirección opuesta. Las probabilidades $T$ y $R$ de transmisión y reflexión están dadas por:

<h3 align="center">$T = {4k_1k_2\over(k_1+k_2)^2}\,,\qquad$
$R = \biggl( {k_1-k_2\over k_1+k_2} \biggr)^2.$</h3>

#### Supongamos que tenemos una partícula con una masa igual a la masa del electrón. $m=9.11\times10^{-31}\,$kg y energía de $10\,$eV, al encontrar un escalón de potencial de altura $9\,$eV. Escribe un programa en \texttt{Python} para calcular e imprimir las probabilidades de transmisión y reflexión utilizando las fórmulas anteriores.


In [11]:
#Primero definimos el valor de cada variable que va a ser constante
#Colocamos el valor de h barra en eV
ℏ = 6.58211e-16
#La masa
m = 9.11e-31
#La energia en eV
E = 10
#El potencial en eV
V = 9

In [12]:
#Ahora calculamos el valor de k1 y k2
k1 = (mt.sqrt(2*m*E))/ℏ
k2 = (mt.sqrt(2*m*(E-V)))/ℏ

In [19]:
#Por ultimo le pedimos que nos pida si queremos ver las probabilidades
while (True):
    p = str(input("Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: "))
    if (p == ':D'):
        T = (4*k1*k2)/((k1+k2)**2)
        R = ((k1-k2)/(k1+k2))**2
        print(f"La probabilidad de transmitancia es: {T}")
        print(f"La probabilidad de reflectancia es: {R}")
        break
    else:
        print("No quiere las probabilidades??")
        print("Ta bien le doy otra oportunidad")

Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: no
No quiere las probabilidades??
Ta bien le doy otra oportunidad
Coloque ':D' si gusta la probabilidad de transmitancia y reflectancia: :D
La probabilidad de transmitancia es: 0.7301261363877617
La probabilidad de reflectancia es: 0.2698738636122385
